# Error handling

## `try-catch-finally`

- 如果你知道调用的方法`有可能`抛出异常，一定要用 try/catch
  - 调用的 library methods document 中有声明 throws
  - 你的 custom exception
  - 经常出现 checked exception 的情况(读取, 读 DB, server request...)
- `finally` block is `optional`, it is `executed regardless of whether an exception is thrown or not`.
  - use case: Cleaning up resource...

```java
// catch blocks must be ordered from most specific to most general(顺序可能会有不同影响)
// Group together similar types of exceptions that are treated the same way using the `|(pipe)`
try {
  [create_code_block]
  [risky_code_block]
} catch ([more_specific_exception | similar_exception]} {   //
  [handling_code_block]
} catch ([more_general_exception]} {
  [handling_code_block2]
} finally {
  [cleanup_code_block]
}
```

- `nested try block`(try block inside a try block) is possible. Sometimes this is unavoidable, but should generally be avoided

## `throws`, `throw`

- `throws` keyword in `method head` declare exceptions that may be thrown **from a method** (tell programmer something may happen)
  - `throws` also allow exception to `propogate the call stack`
- `throw` keyword in `method body` is used to `explicitly throw` an exception

### when to declare an Exception with `throws`? (`Checked exceptions (Exception)` vs `Unchecked exceptions(RuntimeException)`)

- `Checked Exceptions (Exception)`,
  - Exceptions that `out of programmer's control`,
  - required to be thrown using the throws keyword.
  - eg.
    - request an external resource (server may down)
    - read from database (DB may not available)
    - read a file (file may not exist)
- `Unchecked exceptions(RuntimeException)`
  - Unchecked exceptions 大多由程序逻辑错误引起，可以由程序员 test/debug 进行修正。
  - Unchecked exceptions 不需要也不应该使用 throw/throws 声明和抛出, 而是应该尽早发现并修正
  - eg. NullPointerException, NumberFormatException, StackOverflow...

### Ducking❓

- Exception Propagation flow
  - Exception appear (top of call stack)
    - if caught, issue is dealt
    - if not caught, Exception drops down to previous method(`ducking`)... or until it reach the bottom of call stack
- ducking:
  - 在执行风险操作时，不用 try/catch 处理异常，而是用 throws 声明为异常踢皮球给下一级

```java
public void duckRiskyOperation() throws riskyException {
    try {
        RiskyOperation();
    }
}
```

## example: throws, throw, try-catch-finally

```java
// this program throw EvenException when an array contains even number
public class ExceptionDemo {
    public static final int arr[] = {5, 1, 4, 2, 3};
    public static void main(String... args) {
        try {
            checkNotEven(arr);
            System.out.println("Not found, all good!");
        } catch (EvenException e) {
            // handle exception
            System.out.println(e.getMessage());
        } finally {
            System.out.println("Cleaning up arr");
            for (int i = 0; i < arr.length; ++i) {
                arr[i] = 0;
            }
        }
    }
    public static checkNotEven (int... arr) throws EvenException, DummyException {          // throws keyword, may list multiple exceptions
        for (int i = 0; i < arr.length; ++i) {
            if (arr[i]%2 == 0) {
                throw new EvenException("Did not expect an even number") // throw
            }
        }
        return -1;
    }
}
```

## `try with Resources`

- this syntax `eliminate the need to use finally` to manually close the resource. (and the risk of leaking if forget)
- syntax: try([create resource statements])
  - resource created in the try() will be `closed automatically`

```java
static String readFirstLineFromFile(String path) throws IOException {
	    try (FileReader fr = new FileReader(path);
	         BufferedReader br = new BufferedReader(fr)) {
	        return br.readLine();
	    }
	}
```

- may created [Suppressed Exceptions](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) if opening/closing resoures operation within try-with-resource causing exception

## Exceptions

- Exceptions are objects
  - there are some general out-of-box exceptions to use in Java
  - using external library, check their document for library-specific exceptions
  - you can write you custom exceptions

```java
// use a exception:
    // new MyException("my special exception message")
throw new EvenException("Did not expect an even number")
```

### Exceptions are polymorphism

- 例如，AppleException 和 PearException 继承了 FruitException

```java
// throws 声明的 FruitException 允许抛出子类的exception
public void checkFruit() throws FruitException {
    if (appleIssue){
        throw AppleException;
    }
        if (pearIssue){
        throw PearException;
    }
}

// catch FruitException 可以抓到所有子类的exception
try {
    checkFruit()
} catch(FruitException ex) {
    if (ex instanceof AppleException) {
        //fix apple
    }
    if (ex instanceof PearException) {
        //fix pear
    }
}
```

- 因此在有多个 catch 时 spceific(子类)的处理应该在 general(父类)之前，以免发生重复

### custom exception (for your logic and specific need)

- create Unchecked Exception
  - extend the `java.lang.RuntimeException` class

```java
public class IncorrectFileExtensionException
  extends RuntimeException {
    public IncorrectFileExtensionException(String errorMessage, Throwable err) {
        super(errorMessage, err);
    }
}
```

- create checked Exception
  - extend the `Exception` class

```java
// exception
public class IncorrectFileNameException extends Exception {
    public IncorrectFileNameException(String errorMessage) {
        super(errorMessage);
    }
}
```
